flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java

```diff
@@ -557,11 +566,96 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
             return Collections.emptyList();
         }
 
-        List<JavaScriptInvocation> currentList = pendingJsInvocations;
+        List<JavaScriptInvocation> readyToSend = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().map(StateNode::isVisible).orElse(true))
+                .collect(Collectors.toList());
+
+        pendingJsInvocations = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().map(node -> !node.isVisible()).orElse(false))
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        pendingJsInvocations
+                .forEach(this::registerDetachListenerForPendingInvocation);
+
+        return readyToSend;
+    }
+
+    /**
+     *  Registers a detach listener for the given invocation.
+     *
+     * @param invocation - invocation should be always having an existing owner if we call with this.
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private void registerDetachListenerForPendingInvocation(
+            JavaScriptInvocation invocation) {
 
-        pendingJsInvocations = new ArrayList<>();
+        StateNode ownerNode = invocation.getOwner().orElse(ui.getInternals().getStateTree().getRootNode());
+
+        JavaScriptInvocationDetachListener listener = pendingJsInvocationDetachListeners
+                .computeIfAbsent(ownerNode, node -> {
+                    JavaScriptInvocationDetachListener detachListener = new JavaScriptInvocationDetachListener();
+                    detachListener.registration = combine(
+                            () -> pendingJsInvocationDetachListeners
+                                    .remove(node),
+                            node.addDetachListener(detachListener));
+                    return detachListener;
```

Comment:
I would remove the combine method, as it introduces a lot of overhead for a single call (e.g, creating Lists, as it expects potentially more than one registration)
Instead, I would simply create a lambda here
```suggestion
                    Registration registration = node.addDetachListener(detachListener);
                    detachListener.registration = () -> {
                        pendingJsInvocationDetachListeners.remove(node);
                        registration.remove();
                    };
```

(Sorry for duplicate, I messed up with comments)

Reply:
Thank you vm, makes sense, cleaner! Fixed it, but I could not accept the suggestion (will squash all last commits also no worries), because  `return detachListener;` was needed still.